src/libraries/Microsoft.Extensions.Caching.Abstractions/src/Hybrid/HybridCache.cs

```diff
@@ -0,0 +1,128 @@
+    public ValueTask<T> GetOrCreateAsync<T>(string key, Func<CancellationToken, ValueTask<T>> factory,
+        HybridCacheEntryOptions? options = null, IEnumerable<string>? tags = null, CancellationToken cancellationToken = default)
+        => GetOrCreateAsync(key, factory, WrappedCallbackCache<T>.Instance, options, tags, cancellationToken);
+
+    private static class WrappedCallbackCache<T> // per-T memoized helper that allows GetOrCreateAsync<T> and GetOrCreateAsync<TState, T> to share an implementation
+    {
+        // for the simple usage scenario (no TState), pack the original callback as the "state", and use a wrapper function that just unrolls and invokes from the state
+        public static readonly Func<Func<CancellationToken, ValueTask<T>>, CancellationToken, ValueTask<T>> Instance = static (callback, ct) => callback(ct);
+    }
+
+    /// <summary>
+    /// Asynchronously sets or overwrites the value associated with the key.
+    /// </summary>
+    /// <typeparam name="T">The type of the data being considered.</typeparam>
+    /// <param name="key">The key of the entry to create.</param>
+    /// <param name="value">The value to assign for this cache entry.</param>
+    /// <param name="options">Additional options for this cache entry.</param>
+    /// <param name="tags">The tags to associate with this cache entry.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> used to propagate notifications that the operation should be canceled.</param>
+    public abstract ValueTask SetAsync<T>(string key, T value, HybridCacheEntryOptions? options = null, IEnumerable<string>? tags = null, CancellationToken cancellationToken = default);
+
+    /// <summary>
+    /// Asynchronously removes the value associated with the key if it exists.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Not ambiguous in context")]
+    public abstract ValueTask RemoveAsync(string key, CancellationToken cancellationToken = default);
+
+    /// <summary>
+    /// Asynchronously removes the value associated with the key if it exists.
+    /// </summary>
+    /// <remarks>Implementors should treat <c>null</c> as empty</remarks>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters", Justification = "Not ambiguous in context")]
+    public virtual ValueTask RemoveAsync(IEnumerable<string> keys, CancellationToken cancellationToken = default)
+    {
+        return keys switch
+        {
+            // for consistency with GetOrCreate/Set: interpret null as "none"
+            null or ICollection<string> { Count: 0 } => default,
+            ICollection<string> { Count: 1 } => RemoveAsync(keys.Single(), cancellationToken),
```

Comment:
Nit: Single is a bit more expensive than First, as Single needs to validate there's nothing after the first element.

Reply:
switched to First